src/main.cpp

```diff
@@ -3068,10 +3048,38 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
         }
         // When we reach this point, we switched to a new tip (stored in pindexNewTip).
 
+        // Remove orphan transactions with cs_main
+        {
+            LOCK(cs_main);
```

Comment:
@sipa Absolutely, I'm planning on doing that after the dust settles here, unless @TheBlueMatt already has it queued up. I'm planning on moving lots of what's currently in CNode to CNodeState, so breaking out the locks is a prerequisite for that.